### PR TITLE
fix(streaming): fall back after oversized websocket messages

### PR DESCRIPTION
**Pull request by dannote**
## Description

OpenAI Responses WebSocket requests are sent as a single `response.create` text message. For large conversations, the upstream endpoint can reject that message with WebSocket close code `1009` (message too big). Retrying the same WebSocket payload cannot succeed and currently leaves callers with a transport error, even though the same request can be streamed through HTTP/SSE.

This PR adds a narrow, provider-declared fallback:

- recognize close code `1009` before any provider output has been emitted;
- replay the request once through the provider's existing HTTP/SSE streaming path;
- never replay after observable output, avoiding duplicate output or side effects;
- mark caller-owned reusable Responses WebSocket sessions as HTTP-only after fallback, so later turns do not repeat the known-failing transport;
- reset `StreamServer` protocol and provider state when switching from WebSocket events to SSE;
- reuse the existing Finch request construction, retry, telemetry, backpressure, and error handling paths.

Both OpenAI Responses and OpenAI Codex opt into the generic fallback through WebSocket request metadata. Other WebSocket providers retain their existing behavior.

This follows the transport strategy used by the official Codex client:

- [Responses WebSocket endpoint](https://github.com/openai/codex/blob/4ef836f883c38ba6d39e6920f335ce6452b7de33/codex-rs/codex-api/src/endpoint/responses_websocket.rs) sends `response.create` over WebSocket and enables `permessage-deflate`.
- [Responses retry orchestration](https://github.com/openai/codex/blob/4ef836f883c38ba6d39e6920f335ce6452b7de33/codex-rs/core/src/responses_retry.rs) switches transports after the WebSocket retry budget is exhausted.
- [Client transport state](https://github.com/openai/codex/blob/4ef836f883c38ba6d39e6920f335ce6452b7de33/codex-rs/core/src/client.rs) keeps fallback state at session scope.
- [Fallback integration tests](https://github.com/openai/codex/blob/4ef836f883c38ba6d39e6920f335ce6452b7de33/codex-rs/core/tests/suite/websocket_fallback.rs) verify HTTP fallback and that it remains sticky across turns.

ReqLLM intentionally implements the smallest corrective subset here: pre-output `1009` fallback and sticky routing. WebSocket compression and incremental `previous_response_id` reuse remain separate concerns.

## Type of Change

- [x] Bug fix (non-breaking change fixing an issue)
- [ ] New feature (non-breaking change adding functionality)
- [ ] Breaking change (fix or feature causing existing functionality to change)
- [ ] Documentation update

## Breaking Changes

None.

## Testing

- [x] Tests pass (`mix test`)
- [x] Quality checks pass (`mix quality`)

Added integration coverage for:

- pre-output close `1009` replaying successfully over HTTP/SSE;
- close `1009` after provider output remaining an error without replay;
- reusable session fallback remaining sticky across subsequent turns;
- OpenAI Codex WebSocket requests declaring HTTP fallback support.

Validation performed:

- full suite: 4,136 passed, 11 skipped, 151 excluded;
- `mix quality`: formatting, warnings-as-errors compilation, Credo, and Dialyzer passed;
- real `openai_codex:gpt-5.6-sol` normal WebSocket stream succeeded;
- locally forced pre-output close `1009` then replayed successfully against the real Codex HTTP/SSE endpoint;
- the same smoke test verified sticky HTTP routing for the reusable session.

The real smoke test refreshed OAuth credentials in memory only and did not persist or log them.

## Checklist

- [x] My code follows the project's style guidelines
- [x] Documentation is unchanged because this is automatic transport recovery with no new public option
- [x] I have added tests that prove my fix works
- [x] All new and existing tests pass
- [x] My commits follow conventional commit format
- [x] I have **NOT** edited `CHANGELOG.md` (it is auto-generated by git_ops)

## Related Issues

No linked issue.

### CODE REVIEW

File: lib/req_llm/streaming/web_socket_client.ex [hunk 1]
```diff
@@ -102,19 +104,33 @@ defmodule ReqLLM.Streaming.WebSocketClient do
...
-              initial_messages: config.initial_messages,
-              close_on_terminal?: false
-            )
-        end
+        session = reusable_session(opts)
+
+        result =
+          case session do
+            nil ->
+              start_owned_session(config, stream_server_pid, opts)
```
Comment (mikehostetler) [inline lib/req_llm/streaming/web_socket_client.ex:115]: [P1] Close an owned session on every terminal path. `start_owned_session/3` returns only the relay result, so the caller loses the session PID. After a code 1009 close, the task can replay over HTTP and exit while the owned `WebSocketSession` process remains alive. A deterministic process scan after the new test file found two such live sessions. Please retain the owned PID and stop that session after fallback or error. Do not stop a caller-supplied reusable session.

File: lib/req_llm/streaming/web_socket_client.ex [hunk 2]
```diff
@@ -198,11 +218,101 @@ defmodule ReqLLM.Streaming.WebSocketClient do
         :ok
 
       {:error, reason} ->
-        safe_http_event(stream_server_pid, {:error, reason})
-        {:error, reason}
+        if message_too_big?(reason) do
+          {:websocket_error, reason, output_observed?}
```
Comment (mikehostetler) [inline lib/req_llm/streaming/web_socket_client.ex:226]: [P1] Preserve this error when HTTP fallback is not enabled. Every code 1009 close returns `{:websocket_error, reason, false}`, but the next function only handles that result when `fallback_transport` is `:http`. The generic clause then returns the tuple without sending an error event. The task exits normally, so `StreamServer` can finalize an empty successful stream. Please add an explicit no-fallback error path and a test for a provider or configuration with no `fallback_transport`.

File: lib/req_llm/providers/openai.ex [hunk 3]
```diff
@@ -791,14 +791,25 @@ defmodule ReqLLM.Providers.OpenAI do
...
+
+      _ ->
+        :http
     end
   end
 
+  defp websocket_session_fell_back?(session) when is_pid(session) do
+    ReqLLM.Streaming.WebSocketSession.http_fallback?(session)
+  catch
+    :exit, _reason -> true
```
Comment (mikehostetler) [inline lib/req_llm/providers/openai.ex]: [P2] Do not treat a dead or unresponsive reusable session as an established HTTP fallback. This catch changes the request to HTTP even though the session never set `http_fallback?`. That hides a caller-owned session lifecycle failure. Please select HTTP only when the live session reports `true`; otherwise keep the WebSocket path or return the session error. The same catch exists in `OpenAICodex`.

**Review by mikehostetler — CHANGES_REQUESTED** [review]
Thanks for the focused fallback change and the useful transport tests. I found three blocking edge cases:

- An owned WebSocket session remains alive after an HTTP fallback or another error path.
- A code 1009 close becomes an empty successful stream when the provider does not enable HTTP fallback.
- A dead caller-owned session silently selects HTTP.

I opened #959 for the unrelated CI test-isolation failure. That race predates this PR and must stay separate from this change.

Local validation passed for the changed WebSocket and Codex tests, the streaming core tests, and the full suite with the CI seed. The current red CI result is the separate #959 race.

File: lib/req_llm/streaming/web_socket_client.ex (hunk 1 above)
Comment (dannote) [reply lib/req_llm/streaming/web_socket_client.ex:115]: Fixed in 8bf3750a. Owned sessions are now closed after every relay terminal path, while caller-owned reusable sessions remain under caller ownership. Added a regression that verifies no owned WebSocketSession remains after HTTP fallback.

File: lib/req_llm/providers/openai.ex (hunk 3 above)
Comment (dannote) [reply lib/req_llm/providers/openai.ex]: Fixed in 8bf3750a for both OpenAI and OpenAICodex. A dead or unresponsive reusable session now reports false for sticky fallback selection, keeping the WebSocket path so the lifecycle failure remains observable. Added direct coverage for both providers.

File: lib/req_llm/streaming/web_socket_client.ex (hunk 2 above)
Comment (dannote) [reply lib/req_llm/streaming/web_socket_client.ex:226]: Fixed in 8bf3750a. Any code 1009 result that does not match the provider-declared HTTP fallback clause now emits the original stream error instead of allowing a normal empty completion. Added a no-fallback provider regression.

**Review by mikehostetler — APPROVED** [review]
Thank you for the focused follow-up. Commit 8bf3750a resolves all three blocking findings and adds direct regression tests for each case. The focused tests, full local suite, merge check, and GitHub CI all pass. This is ready to merge.
